### PR TITLE
Bootstrap based nonuniform embedding (BBNUE) estimator

### PR DESCRIPTION
**Pull request by kahaaga**
A version of the NUE estimator from  _Montalto, A.; Faes, L.; Marinazzo, D. MuTE: A MATLAB toolbox to compare established and novel estimators of the multivariate transfer entropy. PLoS ONE 2014, 9, e109462._

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -12,16 +12,19 @@ Neighborhood = "645ca80c-8b79-4109-87ea-e1f58159d116"
...
-DSP = "0.6.9"
+DSP = "^0.6.9, ^0.7"
 DelayEmbeddings = "2"
 Distances = "0.10"
 Entropies = "1"
 Neighborhood = "0.2"
 Reexport = "0.2, 0.3, 1"
 SpecialFunctions = "0.8, 0.9, 0.10, 1.2, 2"
 StaticArrays = "1"
+TimeseriesSurrogates = "^1.2.4"
```
Comment (Datseris) [inline Project.toml:27]: As a comment, the `^` is not necessary. The default behavior is to use `^`.

File: src/transferentropy/bbnue.jl [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+""" 
+    transferentropy(source, target, [cond], est::BBNUE; q = 0.95, η = 1, 
+        nsurr = 100, uq = 0.95, 
+        include_instantaneous = true, 
+        method_delay = "ac_min", 
+        maxlag::Union{Int, Float64} = 0.05
+        ) → te, js, τs, idxs_source, idxs_target, idxs_cond
+
+Estimate transfer entropy using the [`BBNUE`](@ref) estimator, which uses a bootstrap-based
```
Comment (Datseris) [inline src/transferentropy/bbnue.jl:26]: I vote that all of this docstring is in `BBNUE` instead. "Polluting" the `transferentropy` docstring with so much details about a specific method isn't ideal for a user. The API is clear that `est` is used in `transferentropy`, there is no gain in highlighting here what `transferentropy` does with this specific method. Instead, better to highlight it at `BBNUE`.

**Comment by Datseris** [comment]
@kahaaga I am not familiar with this method and I won't have time to read the paper to get the necessary amount of familiarity. Thus, I cannot review the actual source code but only the documentation. Just letting you know in advance.

File: src/transferentropy/bbnue.jl [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+For significance testing of a variable, `nsurr` circular shift surrogates are generated, 
+and if transfer entropy for the original variables exceeds the `uq`-quantile of that of the 
+surrogate ensemble, then the variable is included.
+
+If `instantaneous` is `true`, then instantaneous interactions are also considered, i.e. effects like 
+`source(t) → target(t)` are allowed.
+
+In this implementation, the maximum lag for each embedding variable is determined using `estimate_delay` 
+from `DelayEmbeddings.jl`. The keywords `method_delay` (default is "ac_min") controls the method
```
Comment (Datseris) [inline src/transferentropy/bbnue.jl:39]: You either have `DelayEmbeddings` or DelayEmbeddings.jl without code inline.

File: src/transferentropy/bbnue.jl [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+## Returns
+
+A 6-tuple is returned, consisting of:
+- `te`: The computed transfer entropy value. If no relevant variables were selected, then `te = 0.0`.
+- `js`: The indices of the selected variables. `js[i]` is the `i`-th entry in the array `[idxs_source..., idxs_target..., idxs_cond...,]`.
+- `τs`: The embedding lags of the selected variables. `τs[i]` corresponds to `js[i]`.
+- `idxs_source`: The indices of the source variables.
+- `idxs_target`: The indices of the target variables.
+- `idxs_cond`: The indices of the conditional variables (empty if `cond` is not given).
```
Comment (Datseris) [inline src/transferentropy/bbnue.jl:64]: Ah, I see. I guess that's a reason to explain this in the docstring of transfer entropy. Okay, then I think the best thing to do is to create a new function alltogether. I think it is not the best idea to use the same name `transferentropy` for such a massively different return signature.

It is probably more intuitive to make a new function and put it in the docs in the same section as `BBNAU`.

File: Project.toml (hunk 1 above)
Comment (kahaaga) [reply Project.toml:27]: Ok, I didn't catch that from the julia docs. Thanks!

**Comment by kahaaga** [comment]
> @kahaaga I am not familiar with this method and I won't have time to read the paper to get the necessary amount of familiarity. Thus, I cannot review the actual source code but only the documentation. Just letting you know in advance.

@Datseris No worries! I don't expect you to learn a new method just for a PR here. You input on the documentation and interface continues to be very valuable.

> Ah, I see. I guess that's a reason to explain this in the docstring of transfer entropy. Okay, then I think the best thing to do is to create a new function alltogether. I think it is not the best idea to use the same name transferentropy for such a massively different return signature.
> It is probably more intuitive to make a new function and put it in the docs in the same section as BBNAU.

I agree. I hesitated a bit by re-using `transferentropy`. Maybe it could be an idea in the future to use some like `optim_transferentropy(source, target, [cond], est)` as an analogue to `transferentropy(source, target, [cond], est)` for estimators that use some optimization procedure?  

There are several other similar methods that optimize the embedding that I'm implementing at the moment. However, I'm not entirely sure if they would all return the same information yet. Thus, I agree, and think I will just make a method-specific function `bbnue(source, target, [cond], est)` instead of `transferentropy(source, target, [cond], est::BBNUE)`.

Thanks for the input!

**Comment by Datseris** [comment]
yeah `bbnue(...)` seems best

File: Project.toml [hunk 5]
```diff
@@ -12,16 +12,19 @@ Neighborhood = "645ca80c-8b79-4109-87ea-e1f58159d116"
 Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
 SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
 StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
+Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
+TimeseriesSurrogates = "c804724b-8c18-5caa-8579-6025a0767c70"
 
 [compat]
-DSP = "0.6.9"
+DSP = "^0.6.9, ^0.7"
```
Comment (Datseris) [inline Project.toml:19]:
```suggestion
DSP = "^0.6.9, 0.7, 1"
```

**Review by Datseris — APPROVED** [review]
Don't forget to increment minor version in Project.toml, as this is an additional feature.

**Pull request merged**